### ISSUE TITLE
Save Vector layer on completion of draw

### ISSUE THREAD

**Issue by NewToStuff** [Viglino/ol-ext-angular#5]
Hi,
I have built on top of this example to add an Editbar to draw shapes into a vector layer. This has been done by adding to the control component example.

I have also added a MapdataService to handle the updating of these geographical features with an aim of sending the updated data as a GeoJSON to another part of the application for storing and later reloading. I think this is nearly there but it is unclear to me how to detect the updated shape can be signaled to the service in order to add to the observable subject which the service handles.

Basically my strategy is to call updateData() in the MapdataService with the appropriate data sending the entire vector layer each time a shape completion or deletion has taken place. What I am unclear upon is how to detect that the vector has been updated and then access its geography. 

There is a stackblitz here with my code example:

https://stackblitz.com/edit/ol-ext-angular-extend?file=src%2Fapp%2Fmap%2Fmapdata.service.ts

Thanks,

Paul

**Comment by Viglino** [comment]
You have to listen to events `addfeature`, `removefeature` and `changefeature` on your source to register changes on its features and add it to a queue to save it on your server...
```javascript
const changes  = {
  add: [],
  del: [],
  upd: []
}

source.on('addfeature', (f) => {
  changes.add.push(f);
});
source.on('removefeature', (f) => {
  changes.del.push(f);
});
source.on('changefeature', (f) => {
  changes.upd.push(f);
});
```
Then check the changes before sending to your server.
PS: 
- you also have to check if an object has been added or modify before delete and so.
- look at examples using WFS-T with openlayers to see how to achieve this...
- you can look at the [undo/redo interaction](https://viglino.github.io/ol-ext/examples/interaction/map.interaction.undoredo.html): undoInteraction._undoStack lets you get all features modifications.

**Comment by NewToStuff** [comment]
Thanks. I will have a good look at that approach and report back. I was hopeful that I could access all the drawn vectors and grab them all at once each time a geometry is drawn (and load them with editable features on start). If I understand you correctly then this is not a valid approach and I need to manage the creation and deletion of each individual shape separately. All the geometries ultimately belong to a single multi-part geometry in my current use case.

If it helps make it clearer I am trying to load a multi-paart geography from a GeoJSON, edit it and then return the edited completeJSON back to the main application. From what I understand to do this I would need to use your logic to build up this Geometry model, write to GeoJSON and then store. 
I vas thinking there might be something like vectorSource.onChange((e) => {GeographySubject.next(e.getGeoJson()}) . If I read you right I am probably "barking up the wrong tree", i.e not following a workable strategy.

**Comment by Viglino** [comment]
You can also send the whole data to your server in bulk when clicking on a save button...

**Comment by NewToStuff** [comment]
Thanks. I have added the Save button to the Stackblitz example in the control section. When clicked the button has access to a mouseEvent object  - The problem I have is that I cannot see the Geometry in that object and I am not sure how to access an object in another way which contains the Geometry. 

In simple terms - If the Geometry changes then I want to send the whole Geometry to an Observable for the rest of the app to subscribe to. To do that I need to pick up automatically that the geometry has changed then send all of it. I can see (or at least I can) how the add/del/upd approach can work but it seems like a lot of code to maintain for simply sending a completed Geometry. I think you are suggesting a simpler way - I just don't quite understand.

Thanks

**Comment by Viglino** [comment]
Just get features collected in your source...
`vector.getSource().getFeatures()`

**Comment by NewToStuff** [comment]
Hi,

That;s a great help. The "penny has dropped". I will explain in case it helps anyone else.         

I tried that initially (vector.getSource().getFeatures()) in the edit.getInteraction('DrawPolygon').on(['drawend'], function(e){........} part of the code. I got confused as I saw an empty array. The piece I did not understand is that the geometry is not added until after onDrawEnd has been completed. The array therefore contains all the geometry up to the last one drawn - This makes perfect sense now.

Putting the getFeatures(0 as part of a function on the Save Button works perfectly. I don't think I tried this as I do not want the user to have to push a button in order to push the saves to the back end.

The remaining part of the jigsaw is whether there is a callback I can hook into which occurs after the vector is updated. It seems like I can use the following for this if it is any use for anyone else. The object mapdataService is a service I use to transmit the data and should be changed as appropriate.

    vector.on('change', function(e){
    console.log('VECTOR change',e);
    console.log('VECTOR change - features',vector.getSource().getFeatures());
    const geoJsonValue = new GeoJSON().writeFeatures(vector.getSource().getFeatures());
    mapdataService.updateData(geoJsonValue);
    });

Many thanks for your help. This can be closed now.

Paul

**Comment by Viglino** [comment]
Hi,
`drawend` event is trigger before the object is created as it lets the user choose what to do with the drawing (accept or not, fill attributes, etc). You have to delay the action using a setTimeout to get it run.
`change` event is fired when a change occurs (when the feature is inserted). That's why it runs for you.
Using a modify interaction or deleting an object will also trigger a `change` and save the data...

**Comment by NewToStuff** [comment]
Thanks. Much appreciated.
